src/Microsoft.AspNetCore.TestHost/ResponseStream.cs

```diff
@@ -93,126 +77,68 @@ public override void SetLength(long value)
-                    Buffer.BlockCopy(_topBuffer.Array, _topBuffer.Offset, buffer, offset, actualCount);
-                    _topBuffer = new ArraySegment<byte>(_topBuffer.Array,
-                        _topBuffer.Offset + actualCount,
-                        _topBuffer.Count - actualCount);
-                    totalRead += actualCount;
-                    offset += actualCount;
-                    count -= actualCount;
-                }
-                while (count > 0 && (_topBuffer.Count > 0 || _bufferedData.Count > 0));
-                // Keep reading while there is more data available and we have more space to put it in.
-                return totalRead;
-            }
-            finally
-            {
-                _readLock.Release();
-            }
-        }
 
-        public async override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
-        {
-            VerifyBuffer(buffer, offset, count, allowEmpty: false);
-            CancellationTokenRegistration registration = cancellationToken.Register(Abort);
-            await _readLock.WaitAsync(cancellationToken);
-            try
-            {
-                int totalRead = 0;
-                do
-                {
-                    // Don't drained buffered data on abort.
-                    CheckAborted();
-                    if (_topBuffer.Count <= 0)
+                    if (result.IsCompleted)
                     {
-                        byte[] topBuffer = null;
-                        while (!_bufferedData.TryDequeue(out topBuffer))
-                        {
-                            if (_complete)
-                            {
-                                CheckAborted();
-                                // Graceful close
-                                return totalRead;
-                            }
-                            await WaitForDataAsync();
-                        }
-                        _topBuffer = new ArraySegment<byte>(topBuffer);
+                        return 0;
                     }
-                    int actualCount = Math.Min(count, _topBuffer.Count);
-                    Buffer.BlockCopy(_topBuffer.Array, _topBuffer.Offset, buffer, offset, actualCount);
-                    _topBuffer = new ArraySegment<byte>(_topBuffer.Array,
-                        _topBuffer.Offset + actualCount,
-                        _topBuffer.Count - actualCount);
-                    totalRead += actualCount;
-                    offset += actualCount;
-                    count -= actualCount;
+
+                    cancellationToken.ThrowIfCancellationRequested();
+                    Debug.Assert(!result.IsCanceled); // It should only be canceled by cancellationToken.
+
+                    // Try again. TODO: I shouldn't need to do this, there wasn't any data.
+                    _pipe.Reader.AdvanceTo(readableBuffer.End, readableBuffer.End);
```

Comment:
Should be fixed by the next pipelines change, correct @davidfowl  @pakrym

Reply:
No, the reader still wakes up for 0 byte reads. The thing that is fixed right now is completing the pipe without advancing, that works

Reply:
https://github.com/dotnet/corefx/issues/27732